src/installer/managed/Microsoft.NET.HostModel/ResourceUpdater.cs

```diff
@@ -352,126 +182,219 @@ public ResourceUpdater AddResource(byte[] data, string lpType, IntPtr lpName)
+                buffer.WriteArray(trailingSectionStart + delta, moveTrailingSectionBuffer, 0, trailingSectionLength);
 
-            return true;
-        }
+                for (int i = resourceSectionIndex + 1; i < _reader.PEHeaders.SectionHeaders.Length; i++)
+                {
+                    int currentSectionBase = sectionBase + Offsets.OneSectionHeaderSize * i;
 
-        private sealed class EnumResourcesErrorInfo
-        {
-            public int hResult;
-            public bool failedToLockResource;
+                    ModifyI32(buffer, currentSectionBase + Offsets.SectionHeader.VirtualAddress,
+                        pointer => pointer + virtualDelta);
+                    ModifyI32(buffer, currentSectionBase + Offsets.SectionHeader.RawPointer,
+                        pointer => pointer + delta);
+                }
+            }
 
-            public void ThrowException()
+            if (rsrcSectionDataSize != rsrcOriginalVirtualSize)
             {
-                if (failedToLockResource)
+                // update size of .rsrc section
+                int resourceSectionBase = sectionBase + Offsets.OneSectionHeaderSize * resourceSectionIndex;
+                WriteI32(buffer, resourceSectionBase + Offsets.SectionHeader.VirtualSize, rsrcSectionDataSize);
+                WriteI32(buffer, resourceSectionBase + Offsets.SectionHeader.RawSize, newSectionSize);
+
+                void PatchRVA(int offset)
                 {
-                    Debug.Assert(hResult == 0);
-                    throw new ResourceNotAvailableException("Failed to lock resource");
+                    ModifyI32(buffer, offset,
+                        pointer => pointer >= trailingSectionVirtualStart ? pointer + virtualDelta : pointer);
                 }
 
-                Debug.Assert(hResult != 0);
-                throw new HResultException(hResult);
+                // fix header
+                if (_reader.PEHeaders.PEHeader.Magic == PEMagic.PE32Plus)
+                {
+                    ModifyI32(buffer, peSignatureOffset + Offsets.PEHeader.PE64InitializedDataSize,
+                        size => size + delta);
+                    ModifyI32(buffer, peSignatureOffset + Offsets.PEHeader.PE64SizeOfImage,
+                        size => size + virtualDelta);
+
+                    if (needsMoveTrailingSections)
+                    {
+                        // fix RVA in DataDirectory
+                        for (int i = 0; i < _reader.PEHeaders.PEHeader.NumberOfRvaAndSizes; i++)
+                            PatchRVA(peSignatureOffset + Offsets.PEHeader.PE64DataDirectories + i * 8);
+                    }
+
+                    // index of ResourceTable is 2 in DataDirectories
+                    WriteI32(buffer, peSignatureOffset + Offsets.PEHeader.PE64DataDirectories + 2 * 8,
+                        rsrcVirtualAddress);
+                    WriteI32(buffer, peSignatureOffset + Offsets.PEHeader.PE64DataDirectories + 2 * 8 + 4,
```

Comment:
nit: I'd prefer sizeof(int) instead of 4, so it is a bit clearer that it is for the size of the VirtualAddress field.

Reply:
I'm going to add Offsets.DataDirectory.VirtualAddressOffset and Offsets.DataDirectory.SizeOffset instead.